core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,529 @@
+        this.sinkBucketToPartitionMap = requireNonNull(sinkBucketToPartitionMap, "sinkBucketToPartitionMap is null");
+        this.sourceExchanges = ImmutableMap.copyOf(requireNonNull(sourceExchanges, "sourceExchanges is null"));
+        this.sourceBucketToPartitionMap = requireNonNull(sourceBucketToPartitionMap, "sourceBucketToPartitionMap is null");
+        this.sourceBucketNodeMap = requireNonNull(sourceBucketNodeMap, "sourceBucketNodeMap is null");
+        this.remainingRetryAttempts = retryAttempts;
+    }
+
+    public StageId getStageId()
+    {
+        return stageExecution.getStageId();
+    }
+
+    public synchronized ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
+    }
+
+    public synchronized void schedule()
+            throws Exception
+    {
+        if (failure != null) {
+            propagateIfPossible(failure, Exception.class);
+            throw new RuntimeException(failure);
+        }
+
+        if (closed) {
+            return;
+        }
+
+        if (isFinished()) {
+            return;
+        }
+
+        if (!blocked.isDone()) {
+            return;
+        }
+
+        if (taskSource == null) {
+            Map<PlanFragmentId, ListenableFuture<List<ExchangeSourceHandle>>> exchangeSourceHandleMap = sourceExchanges.entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getKey, entry -> toListenableFuture(entry.getValue().getSourceHandles())));
+
+            List<ListenableFuture<List<ExchangeSourceHandle>>> blockedFutures = exchangeSourceHandleMap.values().stream()
+                    .filter(future -> !future.isDone())
+                    .collect(toImmutableList());
+
+            if (!blockedFutures.isEmpty()) {
+                blocked = asVoid(allAsList(blockedFutures));
+                return;
+            }
+
+            Multimap<PlanFragmentId, ExchangeSourceHandle> exchangeSources = exchangeSourceHandleMap.entrySet().stream()
+                    .collect(flatteningToImmutableListMultimap(Map.Entry::getKey, entry -> getFutureValue(entry.getValue()).stream()));
+
+            taskSource = taskSourceFactory.create(
+                    session,
+                    stageExecution.getFragment(),
+                    sourceExchanges,
+                    exchangeSources,
+                    stageExecution::recordGetSplitTime,
+                    sourceBucketToPartitionMap,
+                    sourceBucketNodeMap);
+        }
+
+        while (!queuedPartitions.isEmpty() || !taskSource.isFinished()) {
+            while (queuedPartitions.isEmpty() && !taskSource.isFinished()) {
+                List<TaskDescriptor> tasks = taskSource.getMoreTasks();
+                for (TaskDescriptor task : tasks) {
+                    queuedPartitions.add(task.getPartitionId());
+                    partitionToTaskDescriptorMap.put(task.getPartitionId(), task);
+                    sinkExchange.ifPresent(exchange -> {
```

Comment:
When can this be empty? Comment maybe?

Reply:
This can be empty when the results are consumed via a direct exchange. Added a comment on top of the field.